File: apps/widget/src/shared/content.ts
/**
 * TypeScript mirror of packages/py_contracts/py_contracts/content.py —
 * the canonical MessageContent discriminated union. The API serialises
 * message content in exactly this shape; media/file blocks additionally
 * carry a resolved `url` when read back (MinIO presigned / public URL).
 */

export interface TextBlock {
  kind: "text";
  text: string;
}

export interface MediaBlock {
  kind: "media";
  media_type: "image" | "video" | "audio" | "voice" | "file" | "sticker";
  file_id?: string | null;
  /** Resolved download/display URL (added by the API at read time). */
  url?: string | null;
  caption?: string | null;
  mime?: string | null;
  size?: number | null;
  duration_ms?: number | null;
  width?: number | null;
  height?: number | null;
  /** Original filename, when known. */
  name?: string | null;
}

export interface CardButton {
  text: string;
  action: "url" | "postback";
  value: string;
}

export interface ProductCardBlock {
  kind: "product_card";
  title: string;
  subtitle?: string | null;
  image_file_id?: string | null;
  image_url?: string | null;
  price?: string | null;
  currency?: string | null;
  url?: string | null;
  buttons?: CardButton[];
}

export interface QuickButton {
  id: string;
  text: string;
}

export interface QuickButtonsBlock {
  kind: "quick_buttons";
  text: string;
  buttons: QuickButton[];
}

export interface ButtonReplyBlock {
  kind: "button_reply";
  payload: string;
  text: string;
  flow_session_id?: string | null;
}

export interface SystemEventBlock {
  kind: "system_event";
  event: string;
  meta?: Record<string, unknown>;
}

export type ContentBlock =
  | TextBlock
  | MediaBlock
  | ProductCardBlock
  | QuickButtonsBlock
  | ButtonReplyBlock
  | SystemEventBlock
  // Forward-compat: unknown kinds render as nothing rather than crashing.
  | { kind: string; [k: string]: unknown };

export interface MessageContent {
  blocks: ContentBlock[];
}

export type SenderType = "contact" | "member" | "ai_agent" | "automation" | "system";

/** Message as serialised by the widget REST/WS API. */
export interface WireMessage {
  id: string;
  conversation_id?: string | null;
  sender_type: SenderType;
  sender_name?: string | null;
  sender_avatar_url?: string | null;
  content: MessageContent;
  client_msg_id?: string | null;
  created_at: string; // ISO-8601
  seq?: number | null;
  delivery_status?: "pending" | "sent" | "delivered" | "read" | "failed" | null;
}

File: packages/py_contracts/py_contracts/content.py
"""Canonical message content model — the single discriminated union used by
channel adapters (normalize once), the messages table (stored as-is), the
widget protocol, quick replies, and flow send_message nodes.

Extending: add a new block type here first; adapters degrade per their
capability matrix, never per ad-hoc checks.
"""
from __future__ import annotations

from typing import Annotated, Literal, Union
from uuid import UUID

from pydantic import BaseModel, Field


class TextBlock(BaseModel):
    kind: Literal["text"] = "text"
    text: str


class MediaBlock(BaseModel):
    kind: Literal["media"] = "media"
    media_type: Literal["image", "video", "audio", "voice", "file", "sticker"]
    file_id: UUID
    caption: str | None = None
    mime: str | None = None
    size: int | None = None
    duration_ms: int | None = None
    width: int | None = None
    height: int | None = None


class CardButton(BaseModel):
    text: str
    action: Literal["url", "postback"]
    value: str  # url or postback payload


class ProductCardBlock(BaseModel):
    kind: Literal["product_card"] = "product_card"
    title: str
    subtitle: str | None = None
    image_file_id: UUID | None = None
    image_url: str | None = None
    price: str | None = None
    currency: str | None = None
    url: str | None = None
    buttons: list[CardButton] = Field(default_factory=list)


class QuickButton(BaseModel):
    id: str
    text: str


class QuickButtonsBlock(BaseModel):
    kind: Literal["quick_buttons"] = "quick_buttons"
    text: str
    buttons: list[QuickButton]


class ButtonReplyBlock(BaseModel):
    """Inbound: customer tapped a quick button."""

    kind: Literal["button_reply"] = "button_reply"
    payload: str
    text: str
    flow_session_id: UUID | None = None


class TemplateBlock(BaseModel):
    """WhatsApp Cloud API template send."""

    kind: Literal["template"] = "template"
    template_name: str
    language: str
    components: dict = Field(default_factory=dict)
    category: str | None = None


class LocationBlock(BaseModel):
    kind: Literal["location"] = "location"
    latitude: float
    longitude: float
    name: str | None = None
    address: str | None = None


class EmailBlock(BaseModel):
    kind: Literal["email"] = "email"
    subject: str | None = None
    text: str
    html_body_file_id: UUID | None = None
    headers: dict = Field(default_factory=dict)  # message_id / in_reply_to / references[]
    cc: list[str] = Field(default_factory=list)
    bcc: list[str] = Field(default_factory=list)


class SystemEventBlock(BaseModel):
    """Grey timeline chips: assigned / closed / member joined / etc."""

    kind: Literal["system_event"] = "system_event"
    event: str
    meta: dict = Field(default_factory=dict)


ContentBlock = Annotated[
    Union[
        TextBlock,
        MediaBlock,
        ProductCardBlock,
        QuickButtonsBlock,
        ButtonReplyBlock,
        TemplateBlock,
        LocationBlock,
        EmailBlock,
        SystemEventBlock,
    ],
    Field(discriminator="kind"),
]


class MessageContent(BaseModel):
    """A message = ordered list of blocks. Most are single-block."""

    blocks: list[ContentBlock]

    def plain_text(self) -> str:
        parts: list[str] = []
        for b in self.blocks:
            match b:
                case TextBlock():
                    parts.append(b.text)
                case MediaBlock() if b.caption:
                    parts.append(b.caption)
                case ProductCardBlock():
                    parts.append(b.title)
                case QuickButtonsBlock():
                    parts.append(b.text)
                case ButtonReplyBlock():
                    parts.append(b.text)
                case EmailBlock():
                    parts.append(b.subject or b.text[:200])
                case _:
                    pass
        return "\n".join(p for p in parts if p)
